Support oneofs as query parameters

Apparently `oneof` fields can appear as query parameters. Maybe obvious in hindsight.  

```
error[E0609]: no field `children_of_group` on type `ListGroupsRequest`
   --> src/generated/monitoring/v3/src/transport.rs:204:64
    |
204 |         let builder = builder.query(&[("childrenOfGroup", &req.children_of_group)]);
    |                                                                ^^^^^^^^^^^^^^^^^ unknown field
    |
    = note: available fields are: `name`, `page_size`, `page_token`, `filter`

error[E0609]: no field `ancestors_of_group` on type `ListGroupsRequest`
```

The `filter` is a `oneof`: 

https://github.com/googleapis/googleapis/blob/e134bebad9607975d5d97a6d4fd04fb412aae819/google/monitoring/v3/group_service.proto#L125

The generated code for the query parameters is dead wrong:

```rust
        let builder = builder.query(&[("childrenOfGroup", &req.children_of_group)]);
        let builder = builder.query(&[("ancestorsOfGroup", &req.ancestors_of_group)]);
        let builder = builder.query(&[("descendantsOfGroup", &req.descendants_of_group)]);
```

--- a/generator/internal/language/rust.go
+++ b/generator/internal/language/rust.go
@@ -581,12 +581,16 @@ func rustBaseFieldType(f *api.Field, state *api.APIState, modulePath, sourceSpec
 }
 
 func rustAddQueryParameter(f *api.Field) string {
+	if f.IsOneOf {
+		return rustAddQueryParameterOneOf(f)
+	}
+	fieldName := rustToSnake(f.Name)
 	switch f.Typez {
 	case api.ENUM_TYPE:
 		if f.Optional || f.Repeated {
-			return fmt.Sprintf(`let builder = req.%s.iter().fold(builder, |builder, p| builder.query(&[("%s", p.value())]));`, rustToSnake(f.Name), f.JSONName)
+			return fmt.Sprintf(`let builder = req.%s.iter().fold(builder, |builder, p| builder.query(&[("%s", p.value())]));`, fieldName, f.JSONName)
 		}
-		return fmt.Sprintf(`let builder = builder.query(&[("%s", &req.%s.value())]);`, f.JSONName, rustToSnake(f.Name))
+		return fmt.Sprintf(`let builder = builder.query(&[("%s", &req.%s.value())]);`, f.JSONName, fieldName)
 	case api.MESSAGE_TYPE:
 		// Query parameters in nested messages are first converted to a
 		// `serde_json::Value`` and then recursively merged into the request
@@ -594,14 +598,31 @@ func rustAddQueryParameter(f *api.Field) string {
 		// few requests use nested objects as query parameters. Furthermore,
 		// the conversion is skipped if the object field is `None`.`
 		if f.Optional || f.Repeated {
-			return fmt.Sprintf(`let builder = req.%s.as_ref().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, v| { use gax::query_parameter::QueryParameter; v.add(builder, "%s") });`, rustToSnake(f.Name), f.JSONName)
+			return fmt.Sprintf(`let builder = req.%s.as_ref().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, v| { use gax::query_parameter::QueryParameter; v.add(builder, "%s") });`, fieldName, f.JSONName)
 		}
-		return fmt.Sprintf(`let builder = { use gax::query_parameter::QueryParameter; serde_json::to_value(&req.%s).map_err(Error::serde)?.add(builder, "%s") };`, rustToSnake(f.Name), f.JSONName)
+		return fmt.Sprintf(`let builder = { use gax::query_parameter::QueryParameter; serde_json::to_value(&req.%s).map_err(Error::serde)?.add(builder, "%s") };`, fieldName, f.JSONName)
 	default:
 		if f.Optional || f.Repeated {
-			return fmt.Sprintf(`let builder = req.%s.iter().fold(builder, |builder, p| builder.query(&[("%s", p)]));`, rustToSnake(f.Name), f.JSONName)
+			return fmt.Sprintf(`let builder = req.%s.iter().fold(builder, |builder, p| builder.query(&[("%s", p)]));`, fieldName, f.JSONName)
 		}
-		return fmt.Sprintf(`let builder = builder.query(&[("%s", &req.%s)]);`, f.JSONName, rustToSnake(f.Name))
+		return fmt.Sprintf(`let builder = builder.query(&[("%s", &req.%s)]);`, f.JSONName, fieldName)
+	}
+}
+
+func rustAddQueryParameterOneOf(f *api.Field) string {
+	fieldName := rustToSnakeNoMangling(f.Name)
+	switch f.Typez {
+	case api.ENUM_TYPE:
+		return fmt.Sprintf(`let builder = req.get_%s().iter().fold(builder, |builder, p| builder.query(&[("%s", p.value())]));`, fieldName, f.JSONName)
+	case api.MESSAGE_TYPE:
+		// Query parameters in nested messages are first converted to a
+		// `serde_json::Value`` and then recursively merged into the request
+		// query. The conversion to `serde_json::Value` is expensive, but very
+		// few requests use nested objects as query parameters. Furthermore,
+		// the conversion is skipped if the object field is `None`.`
+		return fmt.Sprintf(`let builder = req.get_%s().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, p| { use gax::query_parameter::QueryParameter; p.add(builder, "%s") });`, fieldName, f.JSONName)
+	default:
+		return fmt.Sprintf(`let builder = req.get_%s().iter().fold(builder, |builder, p| builder.query(&[("%s", p)]));`, fieldName, f.JSONName)
 	}
 }
 

--- a/generator/internal/language/rust_test.go
+++ b/generator/internal/language/rust_test.go
@@ -1351,6 +1351,105 @@ func TestRust_AsQueryParameter(t *testing.T) {
 	}
 }
 
+func TestRust_OneOfAsQueryParameter(t *testing.T) {
+	options := &api.Message{
+		Name:   "Options",
+		ID:     "..Options",
+		Fields: []*api.Field{},
+	}
+	optionsField := &api.Field{
+		Name:     "options_field",
+		JSONName: "optionsField",
+		Typez:    api.MESSAGE_TYPE,
+		TypezID:  options.ID,
+		IsOneOf:  true,
+	}
+	typeField := &api.Field{
+		Name:     "type",
+		JSONName: "type",
+		Typez:    api.INT32_TYPE,
+		IsOneOf:  true,
+	}
+	singularField := &api.Field{
+		Name:     "singular_field",
+		JSONName: "singularField",
+		Typez:    api.STRING_TYPE,
+		IsOneOf:  true,
+	}
+	repeatedField := &api.Field{
+		Name:     "repeated_field",
+		JSONName: "repeatedField",
+		Typez:    api.STRING_TYPE,
+		Repeated: true,
+		IsOneOf:  true,
+	}
+
+	singularEnumField := &api.Field{
+		Name:     "singular_enum_field",
+		JSONName: "singularEnumField",
+		Typez:    api.ENUM_TYPE,
+		IsOneOf:  true,
+	}
+	repeatedEnumField := &api.Field{
+		Name:     "repeated_enum_field",
+		JSONName: "repeatedEnumField",
+		Typez:    api.ENUM_TYPE,
+		Repeated: true,
+		IsOneOf:  true,
+	}
+
+	singularFieldMaskField := &api.Field{
+		Name:     "singular_field_mask",
+		JSONName: "singularFieldMask",
+		Typez:    api.MESSAGE_TYPE,
+		TypezID:  ".google.protobuf.FieldMask",
+		IsOneOf:  true,
+	}
+
+	fields := []*api.Field{
+		typeField,
+		optionsField,
+		singularField, repeatedField,
+		singularEnumField, repeatedEnumField,
+		singularFieldMaskField,
+	}
+	oneof := &api.OneOf{
+		Name:   "one_of",
+		ID:     "..Request.one_of",
+		Fields: fields,
+	}
+	request := &api.Message{
+		Name:   "TestRequest",
+		ID:     "..TestRequest",
+		Fields: fields,
+		OneOfs: []*api.OneOf{oneof},
+	}
+	model := api.NewTestAPI(
+		[]*api.Message{options, request},
+		[]*api.Enum{},
+		[]*api.Service{})
+	api.CrossReference(model)
+	rustLoadWellKnownTypes(model.State)
+
+	for _, test := range []struct {
+		field *api.Field
+		want  string
+	}{
+		{optionsField, `let builder = req.get_options_field().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, p| { use gax::query_parameter::QueryParameter; p.add(builder, "optionsField") });`},
+		{typeField, `let builder = req.get_type().iter().fold(builder, |builder, p| builder.query(&[("type", p)]));`},
+		{singularField, `let builder = req.get_singular_field().iter().fold(builder, |builder, p| builder.query(&[("singularField", p)]));`},
+		{repeatedField, `let builder = req.get_repeated_field().iter().fold(builder, |builder, p| builder.query(&[("repeatedField", p)]));`},
+		{singularEnumField, `let builder = req.get_singular_enum_field().iter().fold(builder, |builder, p| builder.query(&[("singularEnumField", p.value())]));`},
+		{repeatedEnumField, `let builder = req.get_repeated_enum_field().iter().fold(builder, |builder, p| builder.query(&[("repeatedEnumField", p.value())]));`},
+		{singularFieldMaskField, `let builder = req.get_singular_field_mask().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, p| { use gax::query_parameter::QueryParameter; p.add(builder, "singularFieldMask") });`},
+	} {
+		got := rustAddQueryParameter(test.field)
+		if test.want != got {
+			t.Errorf("mismatched as query parameter for %s\nwant=%s\n got=%s", test.field.Name, test.want, got)
+		}
+	}
+}
+
 type rustCaseConvertTest struct {
 	Input    string
 	Expected string

--- a/generator/internal/language/rusttemplate_test.go
+++ b/generator/internal/language/rusttemplate_test.go
@@ -129,7 +129,7 @@ func Test_OneOfAnnotations(t *testing.T) {
 		},
 		FieldType:          "std::string::String",
 		PrimitiveFieldType: "std::string::String",
-		AddQueryParameter:  `let builder = builder.query(&[("oneofField", &req.oneof_field)]);`,
+		AddQueryParameter:  `let builder = req.get_oneof_field().iter().fold(builder, |builder, p| builder.query(&[("oneofField", p)]));`,
 		KeyType:            "",
 		ValueType:          "",
 	}, singular.Codec, ignore); diff != "" {
@@ -147,7 +147,7 @@ func Test_OneOfAnnotations(t *testing.T) {
 		},
 		FieldType:          "std::vec::Vec<std::string::String>",
 		PrimitiveFieldType: "std::string::String",
-		AddQueryParameter:  `let builder = req.oneof_field_repeated.iter().fold(builder, |builder, p| builder.query(&[("oneofFieldRepeated", p)]));`,
+		AddQueryParameter:  `let builder = req.get_oneof_field_repeated().iter().fold(builder, |builder, p| builder.query(&[("oneofFieldRepeated", p)]));`,
 		KeyType:            "",
 		ValueType:          "",
 	}, repeated.Codec, ignore); diff != "" {
@@ -165,7 +165,7 @@ func Test_OneOfAnnotations(t *testing.T) {
 		},
 		FieldType:          "std::collections::HashMap<i32,i32>",
 		PrimitiveFieldType: "std::collections::HashMap<i32,i32>",
-		AddQueryParameter:  `let builder = { use gax::query_parameter::QueryParameter; serde_json::to_value(&req.oneof_field_map).map_err(Error::serde)?.add(builder, "oneofFieldMap") };`,
+		AddQueryParameter:  `let builder = req.get_oneof_field_map().map(|p| serde_json::to_value(p).map_err(Error::serde) ).transpose()?.into_iter().fold(builder, |builder, p| { use gax::query_parameter::QueryParameter; p.add(builder, "oneofFieldMap") });`,
 		KeyType:            "i32",
 		ValueType:          "i32",
 	}, map_field.Codec, ignore); diff != "" {

